src/catalog/src/remote/manager.rs

```diff
@@ -332,6 +201,203 @@ impl RemoteCatalogManager {
+        }));
+    }
+
+    let join_res = futures::future::join_all(joins).await;
+    for item in join_res {
+        let table_id = item.context(ParallelOpenTableSnafu)??;
+        *max_table_id = (*max_table_id).max(table_id);
+    }
+
+    Ok(())
+}
+
+/// Iterate over all table entries on metasrv
+async fn iter_remote_tables<'a>(
+    node_id: u64,
+    backend: &'a KvBackendRef,
+    catalog_name: &'a str,
+    schema_name: &'a str,
+) -> Pin<Box<dyn Stream<Item = Result<(TableGlobalKey, TableGlobalValue)>> + Send + 'a>> {
+    let table_prefix = build_table_global_prefix(catalog_name, schema_name);
+    let mut tables = backend.range(table_prefix.as_bytes());
+    Box::pin(stream!({
+        while let Some(r) = tables.next().await {
+            let Kv(k, v) = r?;
+            if !k.starts_with(table_prefix.as_bytes()) {
+                debug!("Ignoring non-table prefix: {}", String::from_utf8_lossy(&k));
+                continue;
+            }
+            let table_key = TableGlobalKey::parse(&String::from_utf8_lossy(&k))
+                .context(InvalidCatalogValueSnafu)?;
+            let table_value = TableGlobalValue::from_bytes(&v).context(InvalidCatalogValueSnafu)?;
+
+            info!(
+                "Found catalog table entry, key: {}, value: {:?}",
+                table_key, table_value
+            );
+            // metasrv has allocated region ids to current datanode
+            if table_value
+                .regions_id_map
+                .get(&node_id)
+                .map(|v| !v.is_empty())
+                .unwrap_or(false)
+            {
+                yield Ok((table_key, table_value))
+            }
+        }
+    }))
+}
+
+/// Initiates all tables inside a catalog by fetching data from metasrv.
+async fn initiate_tables(
+    node_id: u64,
+    backend: KvBackendRef,
+    engine_manager: TableEngineManagerRef,
+    catalog_name: &str,
+    schema_name: &str,
+    schema: SchemaProviderRef,
+    max_table_id: &mut TableId,
+) -> Result<()> {
+    info!("initializing tables in {}.{}", catalog_name, schema_name);
+    let mut table_num = 0;
+    let tables = iter_remote_tables(node_id, &backend, catalog_name, schema_name).await;
+
+    let kvs = tables.try_collect::<Vec<_>>().await?;
+    let joins = kvs
+        .into_iter()
+        .map(|(table_key, table_value)| {
+            let engine_manager = engine_manager.clone();
+            common_runtime::spawn_bg(async move {
+                open_or_create_table(node_id, engine_manager, &table_key, &table_value).await
+            })
+        })
+        .collect::<Vec<_>>();
+    let vec = futures::future::join_all(joins).await;
+    for res in vec {
+        let table_ref = res.context(ParallelOpenTableSnafu)??;
+        let table_info = table_ref.table_info();
+        let table_name = &table_info.name;
+        let table_id = table_info.ident.table_id;
+        schema.register_table(table_name.clone(), table_ref).await?;
```

Comment:
Question: should we also register tables in parallel?

Reply:
Is `register_table` thread-safe?

Reply:
`register_table` was originally right after `open_or_create_table`, but it used to have `common_runtime::block_on_read`, so it was left serialized.

Now it seems parallelizable to me. Better off, we can try to put it back to near `open_or_create_table`.

cc @v0y4g3r please take a look too